src/Util/ErrorFormatGuesser.php

```diff
@@ -37,14 +37,35 @@ private function __construct()
     public static function guessErrorFormat(Request $request, array $errorFormats): array
     {
         $requestFormat = $request->getRequestFormat('');
-        if ('' !== $requestFormat && isset($errorFormats[$requestFormat])) {
+
+        if ('' !== $requestFormat && array_key_exists($requestFormat, $errorFormats)) {
             return ['key' => $requestFormat, 'value' => $errorFormats[$requestFormat]];
         }
 
-        foreach ($errorFormats as $key => $value) {
-            return ['key' => $key, 'value' => $value];
+        $mimeType = self::getMimeType($request);
+
+        foreach ($errorFormats as $format => $mimeTypes) {
+            if (in_array($mimeType, $mimeTypes, true)) {
+                return ['key' => $format, 'value' => $mimeTypes];
+            }
+        }
+
+        return ['key' => key($errorFormats), 'value' => reset($errorFormats)];
+    }
+
+    /**
+     * Get MIME type from the request.
+     *
+     * @param \Symfony\Component\HttpFoundation\Request $request
+     *
+     * @return string|null
+     */
+    private static function getMimeType(Request $request): ?string
```

Comment:
You can remove this method and just call `$request->getMimeTypes($request->getRequestFormat())` because the content negotiation is already properly handled here: https://github.com/api-platform/core/blob/master/src/EventListener/AddFormatListener.php#L67